ecs_pod_creator.go

```diff
@@ -236,6 +326,253 @@ func (e *EnvironmentVariable) SetSecretOptions(opts SecretOptions) *EnvironmentV
+	o.SupportsDebugMode = &supported
+	return o
+}
+
+// SetTags sets the tags for the pod itself when it is run. This overwrites any
+// existing tags.
+func (o *ECSPodExecutionOptions) SetTags(tags []string) *ECSPodExecutionOptions {
+	o.Tags = tags
+	return o
+}
+
+// AddTags adds new tags to the existing ones for the pod itself when it is run.
+func (o *ECSPodExecutionOptions) AddTags(tags ...string) *ECSPodExecutionOptions {
+	o.Tags = append(o.Tags, tags...)
+	return o
+}
+
+// Validate checks that the placement options are valid.
+func (o *ECSPodExecutionOptions) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	if o.PlacementOpts != nil {
+		catcher.Wrap(o.PlacementOpts.Validate(), "invalid placement options")
+	}
+	if catcher.HasErrors() {
+		return catcher.Resolve()
+	}
+
+	if o.PlacementOpts == nil {
+		o.PlacementOpts = NewECSPodPlacementOptions().SetStrategy(BinpackPlacement).SetStrategyParameter(BinpackMemory)
+	}
+
+	return nil
+}
+
+// ECSPodPlacementOptions represent options to control how an ECS pod is
+// assigned to a container instance.
+type ECSPodPlacementOptions struct {
+	// Strategy is the overall placement strategy. By default, it uses the
+	// binpack strategy.
+	Strategy *ECSPlacementStrategy
+
+	// StrategyParameter is the parameter that determines how the placement
+	// strategy optimizes pod placement. The default value depends on the
+	// strategy:
+	// If the strategy is spread, it defaults to "host".
+	// If the strategy is binpack, it defaults to "memory".
+	// If the strategy is random, this does not apply.
+	StrategyParameter *ECSStrategyParameter
+}
+
+// NewECSPodPlacementOptions creates new options to specify how an ECS pod
+// should be assigned to a container instance.
+func NewECSPodPlacementOptions() *ECSPodPlacementOptions {
+	return &ECSPodPlacementOptions{}
+}
+
+// SetStrategy sets the strategy for placing the pod on a container instance.
+func (o *ECSPodPlacementOptions) SetStrategy(s ECSPlacementStrategy) *ECSPodPlacementOptions {
+	o.Strategy = &s
+	return o
+}
+
+// SetStrategyParameter sets the parameter to optimize for when placing the pod
+// on a container instance.
+func (o *ECSPodPlacementOptions) SetStrategyParameter(p ECSStrategyParameter) *ECSPodPlacementOptions {
+	o.StrategyParameter = &p
+	return o
+}
+
+// Validate checks that the the strategy and its parameter to optimize are a
+// valid combination.
+func (o *ECSPodPlacementOptions) Validate() error {
+	strategy := utility.FromStringPtr(o.Strategy)
+	param := utility.FromStringPtr(o.StrategyParameter)
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(strategy == "" && param == "", "must specify either a strategy, a strategy parameter, or both")
+	if strategy != "" {
+		catcher.ErrorfWhen(!utility.StringSliceContains([]string{SpreadPlacement, RandomPlacement, BinpackPlacement}, strategy), "unrecognized strategy '%s'", strategy)
```

Comment:
Can we create a separate validation function on `ECSPlacementStrategy` for this?

Reply:
Sure, done.